File: swift-bridge/Sources/CoreAnimationBridge/CAEDRMetadata.swift
import Foundation
import QuartzCore

private func caData(_ bytes: UnsafeRawPointer?, _ length: Int) -> Data? {
    guard let bytes, length > 0 else { return nil }
    return Data(bytes: bytes, count: length)
}

@_cdecl("ca_edr_metadata_is_available")
public func ca_edr_metadata_is_available() -> Bool {
    if #available(macOS 13.0, *) {
        return CAEDRMetadata.isAvailable
    }
    return false
}

@_cdecl("ca_edr_metadata_new_hdr10_with_display_info")
public func ca_edr_metadata_new_hdr10_with_display_info(
    _ displayInfoBytes: UnsafeRawPointer?,
    _ displayInfoLen: Int,
    _ contentInfoBytes: UnsafeRawPointer?,
    _ contentInfoLen: Int,
    _ opticalOutputScale: Float
) -> UnsafeMutableRawPointer? {
    guard #available(macOS 10.15, *) else { return nil }
    return caRetain(
        CAEDRMetadata.hdr10(
            displayInfo: caData(displayInfoBytes, displayInfoLen),
            contentInfo: caData(contentInfoBytes, contentInfoLen),
            opticalOutputScale: opticalOutputScale
        )
    )
}

@_cdecl("ca_edr_metadata_new_hdr10")
public func ca_edr_metadata_new_hdr10(
    _ minLuminance: Float,
    _ maxLuminance: Float,
    _ opticalOutputScale: Float
) -> UnsafeMutableRawPointer? {
    guard #available(macOS 10.15, *) else { return nil }
    return caRetain(
        CAEDRMetadata.hdr10(
            minLuminance: minLuminance,
            maxLuminance: maxLuminance,
            opticalOutputScale: opticalOutputScale
        )
    )
}

@_cdecl("ca_edr_metadata_new_hlg")
public func ca_edr_metadata_new_hlg(
    _ ambientViewingEnvironmentBytes: UnsafeRawPointer?,
    _ ambientViewingEnvironmentLen: Int
) -> UnsafeMutableRawPointer? {
    guard #available(macOS 14.0, *), let data = caData(ambientViewingEnvironmentBytes, ambientViewingEnvironmentLen) else {
        return nil
    }
    return caRetain(CAEDRMetadata.hlg(ambientViewingEnvironment: data))
}

@_cdecl("ca_edr_metadata_get_hlg")
public func ca_edr_metadata_get_hlg() -> UnsafeMutableRawPointer? {
    guard #available(macOS 13.0, *) else { return nil }
    return caRetain(CAEDRMetadata.hlg)
}

File: src/ca_edr_metadata.rs
use crate::private::handle_type;

handle_type!(EDRMetadata);

impl EDRMetadata {
    #[must_use]
    pub fn is_available() -> bool {
        unsafe { crate::ffi::ca_edr_metadata_is_available() }
    }

    #[must_use]
    pub fn hdr10_with_display_info(
        display_info: Option<&[u8]>,
        content_info: Option<&[u8]>,
        optical_output_scale: f32,
    ) -> Option<Self> {
        unsafe {
            Self::from_raw(crate::ffi::ca_edr_metadata_new_hdr10_with_display_info(
                display_info.map_or(core::ptr::null(), |value| value.as_ptr().cast()),
                display_info.map_or(0, <[u8]>::len),
                content_info.map_or(core::ptr::null(), |value| value.as_ptr().cast()),
                content_info.map_or(0, <[u8]>::len),
                optical_output_scale,
            ))
        }
    }

    #[must_use]
    pub fn hdr10(
        min_luminance: f32,
        max_luminance: f32,
        optical_output_scale: f32,
    ) -> Option<Self> {
        unsafe {
            Self::from_raw(crate::ffi::ca_edr_metadata_new_hdr10(
                min_luminance,
                max_luminance,
                optical_output_scale,
            ))
        }
    }

    #[must_use]
    pub fn hlg(ambient_viewing_environment: &[u8]) -> Option<Self> {
        unsafe {
            Self::from_raw(crate::ffi::ca_edr_metadata_new_hlg(
                ambient_viewing_environment.as_ptr().cast(),
                ambient_viewing_environment.len(),
            ))
        }
    }

    #[must_use]
    pub fn default_hlg() -> Option<Self> {
        unsafe { Self::from_raw(crate::ffi::ca_edr_metadata_get_hlg()) }
    }
}
